### PR TITLE
[quant][graphmode][fx] Support quantization for standalone module

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45539 [quant] Add support for qat Sigmoid module
* #45538 [quant] Add FixedQParamsFakeQuantize module
* #45473 [quant][graphmode][fx][fix] Fix observer insert logic for ops that inherits quantization parameters for input
* #45292 [quant][graphmode][fx] Merge all quantization mode
* #45343 [quant] Use PlaceholderObserver as default dynamic quant observer
* **#44074 [quant][graphmode][fx] Support quantization for standalone module**

Summary:
Sometimes user need to quantize a submodule as one unit, and this submodule
will be lowered to a different backend like accelerator.

The submodule will be quantized with the same fx based graph mode quantization functions
and will be connected with the rest of the model automatically.

APIs:
```python
class StandaloneModule(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.conv = torch.nn.Conv2d(1, 1, 1)

            def forward(self, x):
                return self.conv(x)

class CustomTracer(Tracer):
      def is_leaf_module(self, m, module_qualified_name):
          return (m.__module__.startswith('torch.nn') and
                     not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, StandaloneModule)

class ModelThatUsesStandaloneModule(...):
      def __init__(self):
          super().__init__()
          self.standalone = StandaloneModule()

      def forward(self, x):
          return self.standalone(x)

m = ModelThatUsesStandaloneModule()
qconfig_dict = {"": qconfig, "standalone_module_name": ["standalone"]}
m = prepare_fx(m, qconfig_dict)
calibrate(m, data)
m = convert_fx(m)

m.standalone = lower_to_acclerator(m.standalone)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23580642](https://our.internmc.facebook.com/intern/diff/D23580642)

### CODE REVIEW

File: torch/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -507,6 +507,20 @@ def convert(self, quantizer, node):
             quantizer.quantized_graph,
             node, quantizer.activation_post_process_map[node.name])
 
+class CustomModule(QuantizeHandler):
+    def convert(self, quantizer, node, load_arg, debug=False):
+        assert node.op == 'call_module'
+        if quantizer.is_dynamic_quant:
+            convert = torch.qauntization.convert_dynamic_child_module_fx
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:514]: nit: `quantization`?

**Comment by vkuzo** [comment]
any chance we can add the proposed example usage to the PR summary?  It would make it easier to review

**Comment by jerryzh168** [comment]
this is not needed due to https://github.com/pytorch/pytorch/pull/44147

**Comment by jerryzh168** [comment]
> any chance we can add the proposed example usage to the PR summary? It would make it easier to review

fyi examples can be found in the test

**Comment by jerryzh168** [comment]
btw, I'm looking for the alternative name of `traceable_custom_module`, looks like this name has a bit overlap with `custom_module` which we implemented in other PRs. maybe `accelerator_submodule`?

File: torch/quantization/fx/quantization_patterns.py [hunk 2]
```diff
@@ -531,6 +531,19 @@ def convert(self, quantizer, node, load_arg, debug=False):
         # module attribute like module._QUANTIZED_INPUT_INDEXES
         return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
 
+class TraceableCustomModuleQuantizeHandler(QuantizeHandler):
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:534]: can we add a docblock?

File: torch/quantization/quantize_fx.py [hunk 3]
```diff
@@ -20,13 +20,33 @@ def fuse_fx(graph_module, inplace=False):
...
+def _prepare_fx(graph_module, qconfig_dict, inplace, is_dynamic_quant, is_child_module=False):
     _check_is_graph_module(graph_module)
     graph_module = fuse_fx(graph_module, inplace)
     quantizer = Quantizer()
     prepare = quantizer.prepare_dynamic if is_dynamic_quant else quantizer.prepare
-    prepared = prepare(graph_module, qconfig_dict, inplace=True)
-    return prepared
+    return prepare(graph_module, qconfig_dict, inplace=True, is_child_module=is_child_module)
+
+def prepare_child_module_fx(graph_module, qconfig_dict, inplace=False):
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:30]: Maybe we can name this `custom_child_module` throughout the code, to clarify the intent?  `child_module` could be confusing as it could mean any module which is a child of an another module.

File: torch/quantization/fx/quantize.py [hunk 4]
```diff
@@ -199,7 +206,7 @@ def get_qconfig(module):
             elif node.op == 'call_module':
                 self.qconfig_map[node.name] = get_qconfig(self.modules[node.target])
 
-    def _prepare(self, model, qconfig_dict, inplace, is_dynamic_quant):
+    def _prepare(self, model, qconfig_dict, inplace, is_dynamic_quant, is_child_module):
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:209]: can we document this extra argument here?  It's hard to understand for someone who hasn't seen this PR recently.

**Review by vkuzo — COMMENTED** [review]
Just curious, have we also considered implementing this without introducing an additional set of APIs?  Could there be a way to mark subregions of the graph to be "child_module", and to not follow regular matching patterns for those subregions?

File: torch/quantization/quantize_fx.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/quantize_fx.py:30]: hmm, this can actually work for any child module.

I'm also looking for names for `traceable custom module`, since this overlaps with the previous custom module a bit. I plan to change the name to `accelerator submodule`

**Comment by jerryzh168** [comment]
> Just curious, have we also considered implementing this without introducing an additional set of APIs? Could there be a way to mark subregions of the graph to be "child_module", and to not follow regular matching patterns for those subregions?

Do you mean e.g. assign some attribute to a submodule? how is this easier than call the registration function?

File: torch/quantization/fx/quantize.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:308]: this is an internal API, docs are in prepare_child_module_fx

**Comment by jerryzh168** [comment]
> Just curious, have we also considered implementing this without introducing an additional set of APIs? Could there be a way to mark subregions of the graph to be "child_module", and to not follow regular matching patterns for those subregions?

I think we can also put it in qconfig_dict, would that make sense?

File: torch/quantization/fx/quantize.py (hunk 4 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py:308]: yes, this would help with readability for people who work on this code (not for users of quantization)

Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:308]: I don't want to repeat the docs though, would it help if I use `see prepare_child_module_fx for docs` for is_child_module?

File: torch/fx/graph_module.py [hunk 5]
```diff
@@ -194,10 +194,20 @@ def __reduce__(self):
     def __deepcopy__(self, memo):
         fake_mod = torch.nn.Module()
         fake_mod.__dict__ = copy.deepcopy(self.__dict__)
-        return GraphModule(fake_mod, self.graph)
+        graph_module = GraphModule(fake_mod, self.graph)
```
Comment (jerryzh168) [inline torch/fx/graph_module.py:197]: these are changes from https://github.com/pytorch/pytorch/pull/45182, will rebase on master after it is landed.

File: torch/quantization/quantize_fx.py [hunk 6]
```diff
@@ -20,13 +20,40 @@ def fuse_fx(graph_module, inplace=False):
...
     _check_is_graph_module(graph_module)
     graph_module = fuse_fx(graph_module, inplace)
     quantizer = Quantizer()
     prepare = quantizer.prepare_dynamic if is_dynamic_quant else quantizer.prepare
-    prepared = prepare(graph_module, qconfig_dict, inplace=True)
-    return prepared
+    return prepare(graph_module, qconfig_dict, inplace=True, is_standalone_module=is_standalone_module)
+
+def _prepare_standalone_module_fx(graph_module, qconfig_dict, inplace=False):
+    r""" [Internal use only] Prepare a standalone module, so that it can be used when quantizing the
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:38]: can we mention that "standalone module" means that the hierarchy will be preserved / things will not be inlined?

File: torch/quantization/fx/quantize.py [hunk 7]
```diff
@@ -298,7 +304,7 @@ def get_qconfig(module_name):
                 self.modules[node.target].qconfig = module_qconfig
                 self.qconfig_map[node.name] = module_qconfig
 
-    def _prepare(self, model, qconfig_dict, inplace, is_dynamic_quant):
+    def _prepare(self, model, qconfig_dict, inplace, is_dynamic_quant, is_standalone_module):
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:307]: I understand that this is duplicating documentation, but can we add a quick comment here with what standalone module means?  It would improve readability and remove the need to navigate to a separate file to understand this argument.

File: torch/quantization/quantize_fx.py [hunk 8]
```diff
@@ -20,13 +20,40 @@ def fuse_fx(graph_module, inplace=False):
...
+    parent module. Used in custom module support
+
+    input of the module is quantized in parent module, output of the module
+    is quantized in the standalone module.
+    Returns:
+      model(GraphModule): prepared standalone module with following attributes:
+        _observed_input_idxs(List[Int]): a list of indexs for the graph inputs that
+                                         needs to be observed in parent module
+        _output_is_observed(Bool): a boolean variable indicate whether the output of the
+                                   custom module is observed or not
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:48]: would it be possible to also explain this in the file where the logic is?  I think duplicating the doc is worth not having to jump between different files.

File: torch/quantization/fx/quantize.py [hunk 9]
```diff
@@ -320,22 +326,32 @@ def _prepare(self, model, qconfig_dict, inplace, is_dynamic_quant):
...
-
         env = {}
         observed_graph = Graph()
         observed_node_names_set = set()
 
         def load_arg(a):
             return map_arg(a, lambda node: env[node.name])
 
+        # indexes for the inputs that needs to be observed
+        observed_input_idxs = []
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:347]: would it make sense to either name this `standalone_module_observed_input_idxs`, or to make it generic and also populate it for non standalone modules?

**Review by vkuzo — APPROVED** [review]
lgtm! some optional nit comments inline

File: torch/quantization/fx/quantize.py (hunk 9 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:347]: Sure, will rename to `standalone_module_observed_input_idxs`

File: torch/quantization/fx/standalone_module.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
+from torch.fx import GraphModule
+
+class ObservedStandaloneGraphModule(GraphModule):
```
Comment (jerryzh168) [inline torch/quantization/fx/standalone_module.py:3]: @jamesr66a @zdevito is this what we want?

Comment (jamesr66a) [reply torch/quantization/fx/standalone_module.py:5]: This looks fine to me in the absence of being able to directly customize what `symbolic_trace` returns. I've filed an issue to track that: https://github.com/pytorch/pytorch/issues/45534
